store/store_action.go

```diff
@@ -385,30 +403,87 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 
 		aJSON.InstanceKey = instanceKey
 
-		actionJSONs[i] = aJSON
+		actionJSONs[actionIndex] = aJSON
+		actionIndex++
 	}
 
+	groupingsAssertions := make(map[string]*snapActionJSON)
+
 	// assertions
 	var assertMaxFormats map[string]int
 	if len(toResolve) > 0 {
-		i := len(actionJSONs) - len(toResolve)
 		for grp, ats := range toResolve {
 			aJSON := &snapActionJSON{
 				Action: "fetch-assertions",
 				Key:    string(grp),
 			}
-			aJSON.Assertions = make([]assertAtJSON, len(ats))
+			aJSON.Assertions = make([]interface{}, len(ats))
+			groupingsAssertions[aJSON.Key] = aJSON
+
 			for j, at := range ats {
-				aJSON.Assertions[j].Type = at.Type.Name
-				aJSON.Assertions[j].PrimaryKey = at.PrimaryKey
+				aj := &assertAtJSON{
+					Type:       at.Type.Name,
+					PrimaryKey: at.PrimaryKey,
+				}
 				rev := at.Revision
 				if rev != asserts.RevisionNotKnown {
-					aJSON.Assertions[j].IfNewerThan = &rev
+					aj.IfNewerThan = &rev
 				}
+				aJSON.Assertions[j] = aj
 			}
-			actionJSONs[i] = aJSON
-			i++
+			actionJSONs[actionIndex] = aJSON
+			actionIndex++
 		}
+	}
+
+	if len(toResolveSeq) > 0 {
+		for grp, ats := range toResolveSeq {
+			key := string(grp)
+			// append to existing grouping if applicable
+			aJSON := groupingsAssertions[key]
+			existingGroup := aJSON != nil
+			if !existingGroup {
+				aJSON = &snapActionJSON{
+					Action: "fetch-assertions",
+					Key:    key,
+				}
+				aJSON.Assertions = make([]interface{}, 0, len(ats))
+			}
+			for _, at := range ats {
+				aj := assertSeqAtJSON{
+					Type:        at.Type.Name,
+					SequenceKey: at.SequenceKey,
+				}
+				// for pinned we request the assertion ​by the sequence point <sequence-number>​, i.e.
+				// {"type": "validation-set",
+				//  "sequence-key": ["16", "account-id", "name"],
+				//  "sequence": <sequence-number>}
+				if at.Pinned {
+					if at.Sequence <= 0 {
+						return nil, nil, fmt.Errorf("internal error: sequence not set for pinned sequence %s, %v", at.Type.Name, at.SequenceKey)
+					}
+					aj.Sequence = at.Sequence
+				} else {
+					// for not pinned, if sequence is specified, then
+					// use it for "if-sequence-equal-or-newer-than": <sequence-number>
+					if at.Sequence > 0 {
+						aj.IfSequenceEqualOrNewerThan = &at.Sequence
+					}
```

Comment:
maybe a comment here that in the else case we will get the latest in the sequence